core/services/offchainreporting/OCRContract.go

```diff
@@ -0,0 +1,228 @@
+}
+
+func (ra *OCRContract) SubscribeToNewConfigs(context.Context) (ocrtypes.ContractConfigSubscription, error) {
+	sub := &OCRContractConfigSubscription{
+		make(chan ocrtypes.ContractConfig),
+		ra,
+		sync.Mutex{},
+		false,
+	}
+	connected := ra.logBroadcaster.Register(ra.contractAddress, sub)
+	if !connected {
+		return nil, errors.New("Failed to register with logBroadcaster")
+	}
+
+	return sub, nil
+}
+
+func (ra *OCRContract) LatestConfigDetails(ctx context.Context) (changedInBlock uint64, configDigest ocrtypes.ConfigDigest, err error) {
+	opts := bind.CallOpts{Context: ctx, Pending: false}
+	result, err := ra.contractCaller.LatestConfigDetails(&opts)
+	if err != nil {
+		return 0, configDigest, errors.Wrap(err, "error getting LatestConfigDetails")
+	}
+	return uint64(result.BlockNumber), ocrtypes.BytesToConfigDigest(result.ConfigDigest[:]), err
+}
+
+// Conform OCRContract to LogListener interface
+type OCRContractConfigSubscription struct {
+	ch       chan ocrtypes.ContractConfig
+	ra       *OCRContract
+	mutex    sync.Mutex
+	chClosed bool
+}
+
+func (sub *OCRContractConfigSubscription) OnConnect() {}
+func (sub *OCRContractConfigSubscription) OnDisconnect() {
+	sub.mutex.Lock()
+	defer sub.mutex.Unlock()
+
+	if !sub.chClosed {
+		sub.chClosed = true
+		close(sub.ch)
+	}
+}
+func (sub *OCRContractConfigSubscription) HandleLog(lb eth.LogBroadcast, err error) {
+	topics := lb.Log().RawLog().Topics
+	if len(topics) == 0 {
+		return
+	}
+	switch topics[0] {
+	case OCRContractConfigSet:
+		configSet, err := sub.ra.contractFilterer.ParseConfigSet(lb.Log().RawLog())
+		if err != nil {
+			panic(err)
+		}
+		configSet.Raw = lb.Log().RawLog()
+		cc := confighelper.ContractConfigFromConfigSetEvent(*configSet)
+		sub.ch <- cc
+	default:
+	}
+}
+func (sub *OCRContractConfigSubscription) JobID() *models.ID {
+	jobID := sub.ra.jobID
+	return &jobID
+}
+func (sub *OCRContractConfigSubscription) Configs() <-chan ocrtypes.ContractConfig {
+	return sub.ch
+}
+func (sub *OCRContractConfigSubscription) Close() {
+	sub.ra.logBroadcaster.Unregister(sub.ra.contractAddress, sub)
+}
+
+func (ra *OCRContract) ConfigFromLogs(ctx context.Context, changedInBlock uint64) (c ocrtypes.ContractConfig, err error) {
+	q := ethereum.FilterQuery{
+		FromBlock: big.NewInt(int64(changedInBlock)),
+		ToBlock:   big.NewInt(int64(changedInBlock)),
+		Addresses: []gethCommon.Address{ra.contractAddress},
+		Topics: [][]gethCommon.Hash{
+			{OCRContractConfigSet},
+		},
+	}
+
+	logs, err := ra.ethClient.FilterLogs(ctx, q)
+	if err != nil {
+		return c, err
+	}
+	if len(logs) == 0 {
+		return c, errors.Errorf("ConfigFromLogs: OCRContract with address 0x%x has no logs", ra.contractAddress)
+	}
+
+	latest, err := ra.contractFilterer.ParseConfigSet(logs[len(logs)-1])
+	if err != nil {
+		return c, errors.Wrap(err, "ConfigFromLogs failed to ParseConfigSet")
+	}
+	latest.Raw = logs[len(logs)-1]
```

Comment:
Can we insert a check here that the address of the OCRContract matches the address in the raw log? This is to prevent a malicious geth node from lying about the address, which could cause safety issues.